internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -21,91 +23,389 @@ import (
 // BuildDestinations creates listeners, routers, clusters, and endpointRefs for all destinations
 // and adds them to the proxyState.
 func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *Builder {
+	var lb *ListenerBuilder
 	if b.proxyCfg.GetDynamicConfig() != nil &&
 		b.proxyCfg.DynamicConfig.Mode == pbmesh.ProxyMode_PROXY_MODE_TRANSPARENT {
-		b.addTransparentProxyOutboundListener(b.proxyCfg.DynamicConfig.TransparentProxy.OutboundListenerPort)
+		lb = b.addTransparentProxyOutboundListener(b.proxyCfg.DynamicConfig.TransparentProxy.OutboundListenerPort)
 	}
 
 	for _, destination := range destinations {
+		b.buildDestination(lb, destination)
+	}
+
+	return b
+}
+
+func (b *Builder) buildDestination(
```

Comment:
I've flattened the implicit/explicit branches due them being so similar after my refactor it felt like a bad idea to duplicate so much code to maintain the split.